### PR TITLE
feat(OptimizelyConfig): Add new fields to OptimizelyConfig.

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
The following new public properties are added to OptimizelyConfig:
- sdkKey
- environmentKey
- attributes
- audiences
- events
- experimentRules and deliveryRules to OptimizelyFeature
- audiences to OptimizelyExperiment

## Test plan
All FSC tests OPTIMIZELY_CONFIG_V2 tests should pass.
FSC OptimizelyConfig [link](https://app.travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/236774233) 

## Dependent PR
https://github.com/optimizely/fullstack-sdk-compatibility-suite/pull/526

## Issues
Here's the simple snippet when you will run multiple times, it will give you different result. Maps are randomly picked. 

```
package main

import (
	"fmt"
)

func main() {
	m := map[string]int {
	  "key1": 1,
	  "key6": 6,
	  "key4": 4,
	  "key3": 3,
	  "key7": 7,
	  "key2": 2,					
	}
	
	for key, val := range m {
		fmt.Println(key)
		fmt.Println(val)		
	}
	fmt.Println("Hello, playground")
}
```
https://stackoverflow.com/questions/55925822/why-are-iterations-over-maps-random

### CODE REVIEW

File: pkg/config/datafileprojectconfig/config.go [hunk 1]
```diff
@@ -213,29 +233,45 @@ func NewDatafileProjectConfig(jsonDatafile []byte, logger logging.OptimizelyLogP
 	attributeMap, attributeKeyToIDMap := mappers.MapAttributes(datafile.Attributes)
 	allExperiments := mappers.MergeExperiments(datafile.Experiments, datafile.Groups)
 	groupMap, experimentGroupMap := mappers.MapGroups(datafile.Groups)
-	experimentMap, experimentKeyMap := mappers.MapExperiments(allExperiments, experimentGroupMap)
+	experiments, experimentMap, experimentKeyMap := mappers.MapExperiments(allExperiments, experimentGroupMap)
 
-	rolloutMap := mappers.MapRollouts(datafile.Rollouts)
+	rollouts, rolloutMap := mappers.MapRollouts(datafile.Rollouts)
+	events := []entities.Event{}
```
Comment (msohailhussain) [inline pkg/config/datafileprojectconfig/config.go:239]: move this code to MapEvents?

File: pkg/config/datafileprojectconfig/config.go [hunk 2]
```diff
@@ -213,29 +233,45 @@ func NewDatafileProjectConfig(jsonDatafile []byte, logger logging.OptimizelyLogP
...
+	for _, event := range datafile.Events {
+		events = append(events, entities.Event(event))
+	}
 	eventMap := mappers.MapEvents(datafile.Events)
 	mergedAudiences := append(datafile.TypedAudiences, datafile.Audiences...)
-	featureMap := mappers.MapFeatures(datafile.FeatureFlags, rolloutMap, experimentMap)
+	features, featureMap := mappers.MapFeatures(datafile.FeatureFlags, rolloutMap, experimentMap)
+	attributes := []entities.Attribute{}
+	for _, attribute := range datafile.Attributes {
```
Comment (msohailhussain) [inline pkg/config/datafileprojectconfig/config.go:247]: same here.

File: pkg/config/datafileprojectconfig/mappers/condition_trees.go [hunk 3]
```diff
@@ -28,6 +28,11 @@ import (
 var errEmptyTree = errors.New("empty tree")
 var json = jsoniter.ConfigCompatibleWithStandardLibrary
 
+// GetDefaultOperators returns default conditional operators
+func GetDefaultOperators() []string {
```
Comment (msohailhussain) [inline pkg/config/datafileprojectconfig/mappers/condition_trees.go:32]: can't you directly assign instead of method.

**Review by msohailhussain — CHANGES_REQUESTED** [review]

File: pkg/config/datafileprojectconfig/mappers/audience.go [hunk 4]
```diff
@@ -34,12 +35,15 @@ func MapAudiences(audiences []datafileEntities.Audience) map[string]entities.Aud
 				// @TODO: handle error
 				func() {}() // cheat the linters
```
Comment (The-inside-man) [inline pkg/config/datafileprojectconfig/mappers/audience.go:36]: Should we clean this up maybe, while we are touching the audiences.?

File: pkg/config/optimizely_config.go [hunk 5]
```diff
@@ -67,89 +106,272 @@ type OptimizelyVariable struct {
...
+					} else {
+						// Checks if item is audience id
+						var audienceName = item
+						if audience, ok := audiencesByID[item]; ok {
+							audienceName = audience.Name
+						}
+						// if audience condition is "NOT" then add "NOT" at start.
+						// Otherwise check if there is already audience id in serializedAudience
+						// then append condition between serializedAudience and item
+						if serializedAudience != "" || cond == "NOT" {
```
Comment (The-inside-man) [inline pkg/config/optimizely_config.go:162]: Shot use a constant for the Operands here.

File: pkg/config/optimizely_config.go [hunk 6]
```diff
@@ -67,89 +106,272 @@ type OptimizelyVariable struct {
...
+			}
+		}
+	}
+	return serializedAudience
+}
 
+func evaluateSubAudience(subAudience, serializedAudience, cond *string) {
+	// Checks if sub audience is empty or not
+	if *subAudience != "" {
+		if *serializedAudience != "" || *cond == "NOT" {
```
Comment (The-inside-man) [inline pkg/config/optimizely_config.go:188]: Same comment here for the operands, then we can reuse the constants.

File: pkg/config/optimizely_config.go [hunk 7]
```diff
@@ -67,89 +106,272 @@ type OptimizelyVariable struct {
...
-		for _, experiment := range feature.FeatureExperiments {
-			experimentVariableMap[experiment.Key] = optimizelyVariableMap
+	}
+}
+
+func getExperimentAudiences(experiment entities.Experiment, audiencesByID map[string]entities.Audience) string {
+	if experiment.AudienceConditions == nil {
+		return ""
+	}
+	return getSerializedAudiences(experiment.AudienceConditions, audiencesByID)
```
Comment (The-inside-man) [inline pkg/config/optimizely_config.go:207]: if getSerializedAudiences returns an empty string when experiment.AudienceConditions is nil then we dont need the first check here and can just use the getSerializedAudiences function.

File: pkg/config/datafileprojectconfig/mappers/experiment.go [hunk 8]
```diff
@@ -23,8 +23,9 @@ import (
 )
 
 // MapExperiments maps the raw experiments entities from the datafile to SDK Experiment entities and also returns a map of experiment key to experiment ID
-func MapExperiments(rawExperiments []datafileEntities.Experiment, experimentGroupMap map[string]string) (experimentMap map[string]entities.Experiment, experimentKeyMap map[string]string) {
+func MapExperiments(rawExperiments []datafileEntities.Experiment, experimentGroupMap map[string]string) (experiments []entities.Experiment, experimentMap map[string]entities.Experiment, experimentKeyMap map[string]string) {
```
Comment (dustin-sier) [inline pkg/config/datafileprojectconfig/mappers/experiment.go:26]: Does this cause any backward compatibility issues? This is public facing and is changing the return types.

**Review by The-inside-man — APPROVED** [review]
Overall LGTM just a few suggestions.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Overall looks great. Some changes suggested.

File: pkg/config/datafileprojectconfig/mappers/experiment.go (hunk 8 above)
Comment (yasirfolio3) [reply pkg/config/datafileprojectconfig/mappers/experiment.go:26]: `mappers` package and everything inside it is for internal use only. We use it only to create projectConfig.

File: pkg/config/datafileprojectconfig/config.go [hunk 9]
```diff
@@ -213,29 +233,45 @@ func NewDatafileProjectConfig(jsonDatafile []byte, logger logging.OptimizelyLogP
...
 		sdkKey:               datafile.SDKKey,
 		environmentKey:       datafile.EnvironmentKey,
+		experiments:          experiments,
 		experimentKeyToIDMap: experimentKeyMap,
 		experimentMap:        experimentMap,
 		groupMap:             groupMap,
+		events:               events,
 		eventMap:             eventMap,
+		features:             features,
 		featureMap:           featureMap,
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/datafileprojectconfig/config.go:271]: Have removed arrays for experiments and features. We still need them for `events`, `audiences` and `rollouts` to maintain order. Also keeping their maps help us for methods like `GetAudienceByID`, `GetEventByKey` etc.

File: pkg/config/optimizely_config.go [hunk 10]
```diff
@@ -67,89 +106,272 @@ type OptimizelyVariable struct {
...
+			}
+			switch item := audience.Conditions.(type) {
+			case string:
+				optlyAudience.Conditions = item
+			case interface{}:
+				jsonConditionsString, err := json.Marshal(item)
+				if err == nil {
+					optlyAudience.Conditions = string(jsonConditionsString)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/optimizely_config.go:126]: that is done in `MapAudiences` method. https://github.com/optimizely/go-sdk/pull/322/files#diff-703f65d96a1b87f13ecee7d84e1b1dd98761ca78b5cfb59ed60f7a261a073e8cR26
We provide it with an array which has `typedAudiences` appended before `audiences`.

File: pkg/config/datafileprojectconfig/config.go [hunk 11]
```diff
@@ -213,29 +233,45 @@ func NewDatafileProjectConfig(jsonDatafile []byte, logger logging.OptimizelyLogP
...
+	experiments, experimentMap, experimentKeyMap := mappers.MapExperiments(allExperiments, experimentGroupMap)
 
-	rolloutMap := mappers.MapRollouts(datafile.Rollouts)
+	rollouts, rolloutMap := mappers.MapRollouts(datafile.Rollouts)
+	events := []entities.Event{}
+	for _, event := range datafile.Events {
+		events = append(events, entities.Event(event))
+	}
 	eventMap := mappers.MapEvents(datafile.Events)
 	mergedAudiences := append(datafile.TypedAudiences, datafile.Audiences...)
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/datafileprojectconfig/config.go:244]: https://github.com/optimizely/go-sdk/pull/322#discussion_r699874681

File: pkg/config/optimizely_config.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/optimizely_config.go:126]: Audiences mapping is tested with datafile `similar_exp_keys_expected`.

**Review by dustin-sier — APPROVED** [review]
LGTM as well.

**Review by dustin-sier — CHANGES_REQUESTED** [review]
LGTM baring changes discussed

**Review by dustin-sier — APPROVED** [review]
This will approve it

**Review by jaeopt — APPROVED** [review]
LGTM with nits

**Review by msohailhussain — APPROVED** [review]
lgtm

**Pull request merged**